### PR TITLE
[NUI] Add a GLView guide

### PR DESCRIPTION
**Pull request by dkdk-ryu**
Add a new guide about GLView in NUI.
GLView is a new class through TCSACR-473.

### CODE REVIEW

**Comment by omar-srbd** [comment]
Hello @dkdk-ryu, I hope you are doing well.
I can see that you have added a new file in this PR.
Please update the `toc_all.md` file to add the new file in this PR. Also, update the overview file (/application/dotnet/guides/user-interface/nui/overview.md) of the Natural User Interface (NUI) to add a short description of the feature.

Thank you.

File: docs/application/dotnet/guides/user-interface/nui/glview.md [hunk 1]
```diff
@@ -0,0 +1,314 @@
...
+
+To create a GLView, follow these steps:
+
+1.  Put ColorFormat in the GLView constructor.
+
+    ```csharp
+    glView = new GLView(GLView.ColorFormat.RGBA8888);
+    ```
+
+2. Set the rendering mode. There are Continuous and OnDemand modes. When you set OnDemand, You have to call RenderOnce() method to invoke the GLRenderFrame callback in the render thread.
```
Comment (omar-srbd) [inline docs/application/dotnet/guides/user-interface/nui/glview.md:34]:
```suggestion
2. Set the rendering mode. There are Continuous and OnDemand modes. When you set OnDemand, You have to call RenderOnce() method to invoke the GLRenderFrame callback in the render thread:
```

File: docs/application/dotnet/guides/user-interface/nui/glview.md [hunk 2]
```diff
@@ -0,0 +1,314 @@
...
+
+    ```csharp
+    glView.RenderingMode = Tizen.NUI.GLRenderingMode.Continuous;
+    ```
+
+3. Set graphis configuration. you have to set surface configuration and GLES version.
```
Comment (omar-srbd) [inline docs/application/dotnet/guides/user-interface/nui/glview.md:40]:
```suggestion
3. Set graphics configuration. you have to set surface configuration and GLES version:
```

File: docs/application/dotnet/guides/user-interface/nui/glview.md [hunk 3]
```diff
@@ -0,0 +1,314 @@
...
+    ```csharp
+    glView.SetGraphicsConfig(true, true, 0, Tizen.NUI.GLESVersion.Version20);
+    ```
+
+4. Register GLInit, GLRender, GLTerminate callbacks.
```
Comment (omar-srbd) [inline docs/application/dotnet/guides/user-interface/nui/glview.md:45]:
```suggestion
4. Register GLInit, GLRender, GLTerminate callbacks:
```

File: docs/application/dotnet/guides/user-interface/nui/glview.md [hunk 4]
```diff
@@ -0,0 +1,314 @@
...
+    ```csharp
+    glView.RegisterGLCallbacks(this.InitializeGL, this.RenderFrameGL, this.Terminate.GL);
+    ```
+
+5. Set resize callback. You can get the resized size of the GLView on the render thread.
```
Comment (omar-srbd) [inline docs/application/dotnet/guides/user-interface/nui/glview.md:50]:
```suggestion
5. Set resize callback. You can get the resized size of the GLView on the render thread:
```

File: docs/application/dotnet/guides/user-interface/nui/glview.md [hunk 5]
```diff
@@ -0,0 +1,314 @@
...
+    ```csharp
+    glView.SetResizeCallback(this.ResizeCallback);
+    ```
+
+## Callbacks
+Callbacks are invoked on the GLView render thread.
```
Comment (omar-srbd) [inline docs/application/dotnet/guides/user-interface/nui/glview.md:56]:
```suggestion

Callbacks are invoked on the GLView render thread:

```

File: docs/application/dotnet/guides/user-interface/nui/glview.md [hunk 6]
```diff
@@ -0,0 +1,314 @@
...
+```csharp
+public delegate void GLInitializeDelegate();
+public delegate int GLRenderFrameDelegate();
+public delegate void GLTerminateDelegate();
+public delegate void ViewResizeDelegate(int w, int h);
+```
+
+These snippets just clear the buffer to white.
+
+### Initialize Callback
```
Comment (omar-srbd) [inline docs/application/dotnet/guides/user-interface/nui/glview.md:66]:
```suggestion
### Initialize callback
```

File: docs/application/dotnet/guides/user-interface/nui/glview.md [hunk 7]
```diff
@@ -0,0 +1,314 @@
...
+```csharp
+public void InitializeGL()
+{
+    //Set the ClearColor to white
+    GL.ClearColor(1.0f, 1.0f, 1.0f, 1.0f);
+}
+```
+
+### Render Callback
```
Comment (omar-srbd) [inline docs/application/dotnet/guides/user-interface/nui/glview.md:75]:
```suggestion
### Render callback
```

File: docs/application/dotnet/guides/user-interface/nui/glview.md [hunk 8]
```diff
@@ -0,0 +1,314 @@
...
+    GL.Viewport(0, 0, width, height);
+    //Clear the buffer with the ClearColor.
+    GL.Clear(ClearBufferMask.ColorBufferBit);
+
+    //When returning not 0, the eglSwapBuffers is called.
+    return 1;
+}
+```
+
+### Terminate Callback
```
Comment (omar-srbd) [inline docs/application/dotnet/guides/user-interface/nui/glview.md:89]:
```suggestion
### Terminate callback
```

File: docs/application/dotnet/guides/user-interface/nui/glview.md [hunk 9]
```diff
@@ -0,0 +1,314 @@
...
+```csharp
+public void TerminateGL()
+{
+    // Destroy GL resources
+}
+```
+
+### Resize Callback
```
Comment (omar-srbd) [inline docs/application/dotnet/guides/user-interface/nui/glview.md:97]:
```suggestion
### Resize callback
```

File: docs/application/dotnet/guides/user-interface/nui/glview.md [hunk 10]
```diff
@@ -0,0 +1,314 @@
...
+Table: ImageView control properties
+
+| Property             | Type        | Description                              |
+|--------------------|-----------|----------------------------------------|
+| `RenderingMode`        | `GLRenderingMode`    | The rendering mode of GLView        |
+
+
+
+
+## Triangle Example
```
Comment (omar-srbd) [inline docs/application/dotnet/guides/user-interface/nui/glview.md:118]:
```suggestion
## Triangle example
```

File: docs/application/dotnet/guides/user-interface/nui/glview.md [hunk 11]
```diff
@@ -0,0 +1,314 @@
...
+using Tizen.NUI.Components;
+using OpenTK.Graphics.ES20;
+```
+This guide uses OpenTK.Tizen.
+
+## Create a GLView
+
+To create a GLView, follow these steps:
+
+1.  Put ColorFormat in the GLView constructor.
```
Comment (omar-srbd) [inline docs/application/dotnet/guides/user-interface/nui/glview.md:28]:
```suggestion
1.  Put ColorFormat in the GLView constructor:
```

File: docs/application/dotnet/guides/user-interface/nui/glview.md [hunk 12]
```diff
@@ -0,0 +1,314 @@
...
+### Terminate Callback
+```csharp
+public void TerminateGL()
+{
+    // Destroy GL resources
+}
+```
+
+### Resize Callback
+```csharp
```
Comment (omar-srbd) [inline docs/application/dotnet/guides/user-interface/nui/glview.md:98]:
```suggestion

Resize callback with the following code:

```csharp
```

File: docs/application/dotnet/guides/user-interface/nui/glview.md [hunk 13]
```diff
@@ -0,0 +1,314 @@
...
+    //Clear the buffer with the ClearColor.
+    GL.Clear(ClearBufferMask.ColorBufferBit);
+
+    //When returning not 0, the eglSwapBuffers is called.
+    return 1;
+}
+```
+
+### Terminate Callback
+```csharp
```
Comment (omar-srbd) [inline docs/application/dotnet/guides/user-interface/nui/glview.md:90]:
```suggestion

Terminate callback with the following code:

```csharp
```

File: docs/application/dotnet/guides/user-interface/nui/glview.md [hunk 14]
```diff
@@ -0,0 +1,314 @@
...
+public void InitializeGL()
+{
+    //Set the ClearColor to white
+    GL.ClearColor(1.0f, 1.0f, 1.0f, 1.0f);
+}
+```
+
+### Render Callback
+```csharp
```
Comment (omar-srbd) [inline docs/application/dotnet/guides/user-interface/nui/glview.md:76]:
```suggestion

Render callback by using the following code:

```csharp
```

File: docs/application/dotnet/guides/user-interface/nui/glview.md [hunk 15]
```diff
@@ -0,0 +1,314 @@
...
+public delegate void GLInitializeDelegate();
+public delegate int GLRenderFrameDelegate();
+public delegate void GLTerminateDelegate();
+public delegate void ViewResizeDelegate(int w, int h);
+```
+
+These snippets just clear the buffer to white.
+
+### Initialize Callback
+```csharp
```
Comment (omar-srbd) [inline docs/application/dotnet/guides/user-interface/nui/glview.md:67]:
```suggestion

Initialize callback by using the following code:

```csharp
```

File: docs/application/dotnet/guides/user-interface/nui/glview.md [hunk 16]
```diff
@@ -0,0 +1,314 @@
...
+
+| Property             | Type        | Description                              |
+|--------------------|-----------|----------------------------------------|
+| `RenderingMode`        | `GLRenderingMode`    | The rendering mode of GLView        |
+
+
+
+
+## Triangle Example
+```csharp
```
Comment (omar-srbd) [inline docs/application/dotnet/guides/user-interface/nui/glview.md:119]:
```suggestion

The following code shows how a triangle is rendered using GLView with OpenGL&reg; ES:

```csharp
```

**Review by omar-srbd — COMMENTED** [review]
Hello @dkdk-ryu,
I have reviewed the file in this PR and left some suggestions for you. Please implement them at your earliest convenience. Also, is there any specific reason on why this PR has been kept in **Draft** stage?

**Comment by dkdk-ryu** [comment]
> Hello @dkdk-ryu, I hope you are doing well. I can see that you have added a new file in this PR. Please update the `toc_all.md` file to add the new file in this PR. Also, update the overview file (/application/dotnet/guides/user-interface/nui/overview.md) of the Natural User Interface (NUI) to add a short description of the feature.
> 
> Thank you.

Thank you for the your review.
Speaking of overview.md, there is an issue. refer #1522 .
The author removed CameraView. NUI developers think that CameraView have not to be removed.
As far as I know the issue is related to XAML documentations.
So I should check the status.

File: docs/application/dotnet/guides/user-interface/nui/glview.md [hunk 17]
```diff
@@ -0,0 +1,331 @@
...
+    glView = new GLView(GLView.ColorFormat.RGBA8888);
+    ```
+
+2. Set the rendering mode. There are Continuous and OnDemand modes. When you set OnDemand, You have to call RenderOnce() method to invoke the GLRenderFrame callback in the render thread:
+
+    ```csharp
+    glView.RenderingMode = Tizen.NUI.GLRenderingMode.Continuous;
+    ```
+
+3. Set graphics configuration. you have to set surface configuration and GLES version:
```
Comment (omar-srbd) [inline docs/application/dotnet/guides/user-interface/nui/glview.md:40]:
```suggestion
3. Set graphics configuration. You have to set surface configuration and GLES version:
```

File: docs/application/dotnet/guides/user-interface/nui/glview.md [hunk 18]
```diff
@@ -0,0 +1,331 @@
...
+    ```csharp
+    glView.SetGraphicsConfig(true, true, 0, Tizen.NUI.GLESVersion.Version20);
+    ```
+
+4. Register GLInit, GLRender, GLTerminate callbacks:
+    ```csharp
```
Comment (omar-srbd) [inline docs/application/dotnet/guides/user-interface/nui/glview.md:46]:
```suggestion
    
    ```csharp
```

File: docs/application/dotnet/guides/user-interface/nui/glview.md [hunk 19]
```diff
@@ -0,0 +1,331 @@
...
+    ```
+
+2. Set the rendering mode. There are Continuous and OnDemand modes. When you set OnDemand, You have to call RenderOnce() method to invoke the GLRenderFrame callback in the render thread:
+
+    ```csharp
+    glView.RenderingMode = Tizen.NUI.GLRenderingMode.Continuous;
+    ```
+
+3. Set graphics configuration. you have to set surface configuration and GLES version:
+    ```csharp
```
Comment (omar-srbd) [inline docs/application/dotnet/guides/user-interface/nui/glview.md:41]:
```suggestion
    
    ```csharp
```

File: docs/application/dotnet/guides/user-interface/nui/glview.md [hunk 20]
```diff
@@ -0,0 +1,331 @@
...
+    glView.SetGraphicsConfig(true, true, 0, Tizen.NUI.GLESVersion.Version20);
+    ```
+
+4. Register GLInit, GLRender, GLTerminate callbacks:
+    ```csharp
+    glView.RegisterGLCallbacks(this.InitializeGL, this.RenderFrameGL, this.Terminate.GL);
+    ```
+
+5. Set resize callback. You can get the resized size of the GLView on the render thread:
+    ```csharp
```
Comment (omar-srbd) [inline docs/application/dotnet/guides/user-interface/nui/glview.md:51]:
```suggestion
    
    ```csharp
```

File: docs/application/dotnet/guides/user-interface/nui/glview.md [hunk 21]
```diff
@@ -0,0 +1,331 @@
...
+
+To create a GLView, follow these steps:
+
+1.  Put ColorFormat in the GLView constructor:
+
+    ```csharp
+    glView = new GLView(GLView.ColorFormat.RGBA8888);
+    ```
+
+2. Set the rendering mode. There are Continuous and OnDemand modes. When you set OnDemand, You have to call RenderOnce() method to invoke the GLRenderFrame callback in the render thread:
```
Comment (omar-srbd) [inline docs/application/dotnet/guides/user-interface/nui/glview.md:34]:
```suggestion
2. Set the rendering mode. There are Continuous and OnDemand modes. When you set OnDemand, you have to call RenderOnce() method to invoke the GLRenderFrame callback in the render thread:
```

**Review by omar-srbd — COMMENTED** [review]
Hello, @dkdk-ryu, thanks for updating the document as per the suggestions. After another check of the document, I found a few more issues that need to be addressed.
Please check the suggestions and implement them at your convenience.
Thanks.

File: docs/application/dotnet/guides/user-interface/nui/glview.md [hunk 22]
```diff
@@ -0,0 +1,334 @@
...
+![GLView](./media/GLView.png)
+
+
+## Add namespace
+To use GLView, include `Tizen.NUI.Components`  and `OpenTK.Graphics` namespace in your application. You should refer to the `OpenTK` usage:
+
+```csharp
+using Tizen.NUI;
+using Tizen.NUI.Components;
+using OpenTK.Graphics.ES20;
```
Comment (hyunjushin) [inline docs/application/dotnet/guides/user-interface/nui/glview.md:20]: why don't you use ES30 instead of ES20? ES20 is quite old version.

Comment (dkdk-ryu) [reply docs/application/dotnet/guides/user-interface/nui/glview.md:20]: The reason why I chose ES20 is that it is the mandatory version of OpenGL ES.
I've checked the android page today, It uses GLES2 on the page. https://developer.android.com/reference/android/opengl/GLSurfaceView

If you would like to use GLES3, I will change it.

**Comment by omar-srbd** [comment]
Hello @dkdk-ryu,
The new file (/application/dotnet/guides/user-interface/nui/glview.md) in this PR needs to be added to the overview.md file (/application/dotnet/guides/user-interface/nui/ui-components/).
Have you checked the status regarding Cameraview and other issues in #1522?

**Comment by dkdk-ryu** [comment]
> Hello @dkdk-ryu, The new file (/application/dotnet/guides/user-interface/nui/glview.md) in this PR needs to be added to the overview.md file (/application/dotnet/guides/user-interface/nui/ui-components/). Have you checked the status regarding Cameraview and other issues in #1522?

Sorry, I'm going to check it again.

**Comment by omar-srbd** [comment]
Hello @dkdk-ryu, this PR has been inactive for a while, can you please provide me an update on https://github.com/Samsung/tizen-docs/pull/1732#issuecomment-1333267032?

Thank you.

**Comment by dkdk-ryu** [comment]
I checked #1744 status, and I modified ui-components.md.

**Review by omar-srbd — APPROVED** [review]
@dkdk-ryu, thank you for incorporating all the changes as per the suggestions.

@mijin85cho @asm-foysal I am approving this PR for merging.

**Pull request merged**